contracts/Lens/ComptrollerLens.sol

```diff
@@ -105,113 +126,240 @@ contract ComptrollerLens is ComptrollerLensInterface, ComptrollerErrorReporter,
+     *      It also calculates the health factor and average liquidation threshold.
+     */
+    function _calculateAccountPosition(
+        address comptroller,
+        address account,
+        VToken vTokenModify,
+        uint256 redeemTokens,
+        uint256 borrowAmount,
+        function(address) external view returns (uint256) weight
+    ) internal view returns (uint256 errorCode, AccountLiquidityLocalVars memory vars) {
         // For each asset the account is in
         VToken[] memory assets = ComptrollerInterface(comptroller).getAssetsIn(account);
-        uint assetsCount = assets.length;
-        for (uint i = 0; i < assetsCount; ++i) {
+        uint256 assetsCount = assets.length;
+        ResilientOracleInterface oracle = ComptrollerInterface(comptroller).oracle();
+
+        for (uint256 i = 0; i < assetsCount; ++i) {
             VToken asset = assets[i];
 
             // Read the balances and exchange rate from the vToken
-            (oErr, vars.vTokenBalance, vars.borrowBalance, vars.exchangeRateMantissa) = asset.getAccountSnapshot(
+            (errorCode, vars.vTokenBalance, vars.borrowBalance, vars.exchangeRateMantissa) = asset.getAccountSnapshot(
                 account
             );
-            if (oErr != 0) {
-                // semi-opaque error code, we assume NO_ERROR == 0 is invariant between upgrades
-                return (uint(Error.SNAPSHOT_ERROR), 0, 0);
+            if (errorCode != 0) {
+                revert SnapshotError();
             }
-            (, uint collateralFactorMantissa) = ComptrollerInterface(comptroller).markets(address(asset));
-            vars.collateralFactor = Exp({ mantissa: collateralFactorMantissa });
-            vars.exchangeRate = Exp({ mantissa: vars.exchangeRateMantissa });
 
             // Get the normalized price of the asset
-            vars.oraclePriceMantissa = ComptrollerInterface(comptroller).oracle().getUnderlyingPrice(address(asset));
+            vars.oraclePriceMantissa = oracle.getUnderlyingPrice(address(asset));
             if (vars.oraclePriceMantissa == 0) {
-                return (uint(Error.PRICE_ERROR), 0, 0);
+                revert PriceError(address(asset));
             }
-            vars.oraclePrice = Exp({ mantissa: vars.oraclePriceMantissa });
 
-            // Pre-compute a conversion factor from tokens -> bnb (normalized price value)
-            vars.tokensToDenom = mul_(mul_(vars.collateralFactor, vars.exchangeRate), vars.oraclePrice);
+            Exp memory vTokenPrice = mul_(
+                Exp({ mantissa: vars.exchangeRateMantissa }),
+                Exp({ mantissa: vars.oraclePriceMantissa })
+            );
+            Exp memory weightedVTokenPrice = mul_(Exp({ mantissa: weight(address(asset)) }), vTokenPrice);
+
+            vars.totalCollateral = mul_ScalarTruncateAddUInt(vTokenPrice, vars.vTokenBalance, vars.totalCollateral);
+
+            vars.liquidationThresholdAvg = mul_ScalarTruncateAddUInt(
```

Comment:
I think we don't need to make this update, because this value is already stored in `vars.weightedCollateral`. So, I think we could remove this line here and use `vars.weightedCollateral` in the numerator of the division we have in the line 320, in the function `_finalizeSnapshot`